cdc/processor/sinkmanager/table_sink_advancer.go

```diff
@@ -0,0 +1,354 @@
+		zap.Bool("splitTxn", a.splitTxn),
+		zap.Uint64("currTxnCommitTs", a.currTxnCommitTs),
+		zap.Uint64("lastTxnCommitTs", a.lastTxnCommitTs),
+		zap.Bool("isLastTime", isLastTime))
+
+	// Still got the same commit ts.
+	if a.currTxnCommitTs == a.lastPos.CommitTs {
+		// All transactions before currTxnCommitTs are resolved.
+		if a.lastPos.IsCommitFence() {
+			err = advanceTableSink(a.task, a.currTxnCommitTs,
+				a.committedTxnSize+a.pendingTxnSize, a.sinkMemQuota)
+		} else {
+			// This means all events of the current transaction have been fetched, but we can't
+			// ensure whether there are more transaction with the same CommitTs or not.
+			// So we need to advance the table sink with a batchID. It will make sure that
+			// we do not cross the CommitTs boundary.
+			err = advanceTableSinkWithBatchID(a.task, a.currTxnCommitTs,
+				a.committedTxnSize+a.pendingTxnSize, a.batchID, a.sinkMemQuota)
+			a.batchID += 1
+		}
+
+		a.committedTxnSize = 0
+		a.pendingTxnSize = 0
+	} else if a.splitTxn && a.currTxnCommitTs > 0 {
+		// We just got a new commit ts. Because we split the transaction,
+		// we can advance the table sink with the current commit ts.
+		// This will advance some complete transactions before currTxnCommitTs,
+		// and one partial transaction with `batchID`.
+		err = advanceTableSinkWithBatchID(a.task, a.currTxnCommitTs,
+			a.committedTxnSize+a.pendingTxnSize, a.batchID, a.sinkMemQuota)
+
+		a.batchID += 1
+		a.committedTxnSize = 0
+		a.pendingTxnSize = 0
+	} else if !a.splitTxn && a.lastTxnCommitTs > 0 {
+		// We just got a new commit ts. Because we don't split the transaction,
+		// we **only** advance the table sink by the last transaction commit ts.
+		err = advanceTableSink(a.task, a.lastTxnCommitTs,
+			a.committedTxnSize, a.sinkMemQuota)
+		a.committedTxnSize = 0
+		// If it is the last time we call `advance`, but `pendingTxnSize`
+		// hasn't been recorded yet. To avoid losing it, record it manually.
+		if isLastTime && a.pendingTxnSize > 0 {
+			a.sinkMemQuota.Record(a.task.span,
+				model.NewResolvedTs(a.currTxnCommitTs), a.pendingTxnSize)
+			a.pendingTxnSize = 0
+		}
+	}
+	return
+}
+
+// lastTimeAdvance only happens when there is no enough memory quota to
+// acquire, and the task is not finished.
+// In this case, we need to try to advance the table sink as much as possible.
+func (a *tableSinkAdvancer) lastTimeAdvance() error {
+	return a.advance(true)
+}
+
+// tryAdvanceAndAcquireMem tries to acquire the memory quota and advance the table sink.
+// allFinished indicates whether all the events have been fetched. Then we
+// do not need to acquire the memory quota anymore.
+// txnFinished indicates whether the current transaction has been finished.
+// If it is finished, it is OK to wait next round task to advance the table sink.
+// Otherwise, we need to advance the table at least to the current transaction.
+func (a *tableSinkAdvancer) tryAdvanceAndAcquireMem(
+	allFinished bool,
```

Comment:
Is `allFetched` more accurate according to comments?

Reply:
Updated! Thanks!